clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -420,9 +421,14 @@ boolean joinGroupIfNeeded(final Timer timer) {
                 // need to set the flag before calling onJoinPrepare since the user callback may throw
                 // exception, in which case upon retry we should not retry onJoinPrepare either.
                 needsJoinPrepare = false;
-                onJoinPrepare(generation.generationId, generation.memberId);
+                if (!onJoinPrepare(generation.generationId, generation.memberId))
```

Comment:
This logic is a bit complicated now, thinking about it twice I feel it may read simpler as this:

```
if (needsJoinPrepare) {
    try {
        if (onJoinPrepare(generation.generationId, generation.memberId))
            needsJoinPrepare = false;
        else
            return false;
    } finally {
            // if onJoinPrepare throws an exception, it would be from the rebalance listener.
            // next time we would then not retry {@code onJoinPrepare} any more but proceed the join-group procedure.
            needsJoinPrepare = false;
    }
}
```

Reply:
If you do
This goal should not be achieved:
```
if (!onJoinPrepare(generation.generationId, generation.memberId))
                     needsJoinPrepare = true;
```

Reply:
You're right. How about this:
```java

if (needsJoinPrepare) {
    try {
        if (onJoinPrepare(generation.generationId, generation.memberId))
            needsJoinPrepare = false;
        else
            return false;
    } catch (KafkaException e) {
            // if onJoinPrepare throws an exception, it would be from the rebalance listener.
            // next time we would then not retry {@code onJoinPrepare} any more but proceed the join-group procedure.
            needsJoinPrepare = false;
            throw e;
    }
}
```

Reply:
I think it can be simplified like this
 ```
           if (needsJoinPrepare) {
                // need to set the flag before calling onJoinPrepare since the user callback may throw
                // exception, in which case upon retry we should not retry onJoinPrepare either.
                needsJoinPrepare = false;
                if (!onJoinPrepare(generation.generationId, generation.memberId)) {
                    needsJoinPrepare = true;
                    //should not initiateJoinGroup if needsJoinPrepare still is true
                    return false;
                }
            }
```